Source/WebCore/rendering/RenderLayerCompositor.cpp

```diff
@@ -1311,49 +1321,50 @@ void RenderLayerCompositor::computeCompositingRequirements(RenderLayer* ancestor
     LayerListMutationDetector mutationChecker(layer);
 #endif
 
-    bool anyDescendantHas3DTransform = false;
     bool descendantsAddedToOverlap = currentState.hasNonRootCompositedAncestor();
 
-    if (layer.hasNegativeZOrderLayers()) {
-        // Speculatively push this layer onto the overlap map.
-        bool didSpeculativelyPushOverlapContainer = false;
-        if (!didPushOverlapContainer) {
-            overlapMap.pushSpeculativeCompositingContainer(layer);
-            didPushOverlapContainer = true;
-            didSpeculativelyPushOverlapContainer = true;
-        }
+    if (!canSkipComputeCompositingRequirementsForSubtree(layer, willBeComposited)) {
```

Comment:
It's a bit weird to see this here in the flow that isn't `traverseUnchangedSubtree` I'd ask myself which we didn't go down `traverseUnchangedSubtree` instead, which could itself early return.

That suggests that maybe this fix should be about managing the dirty bits, rather than more branching in this logic.

Reply:
The issue is that the `fullPaintOrderTraversalRequired` and  `descendantsRequireCompositingUpdate` compositing state bits can be set.

In the specific case I was debugging, we changed the height of the page, which caused the RenderView's compositing bounds to change and cause `descendantsRequireCompositingUpdate` to be set (via `setSubsequentLayersNeedCompositingRequirementsTraversal()`).
Also, a later sibling layer got removed, which dirties the z-order lists of the shared parent, and sets the `fullPaintOrderTraversalRequired`. That's maybe avoidable with smart tracking of exactly what did change in the z-order lists.

We could use these same conditions to override the checks about whether to go into `traverseUnchangedSubtree`, and then early return inside it?

I think we need to handle the case where something behind the layer has changed, and we really do need to re-evaluate if the layer should be composited or not. If it does still composite, then we can skip over its descendants. That's the intent of the current interception point.